akka-distributed-data/src/main/scala/akka/cluster/ddata/ORMap.scala

```diff
@@ -174,8 +303,123 @@ final class ORMap[A, B <: ReplicatedData] private[akka] (
         case (None, None) ⇒ throw new IllegalStateException(s"missing value for $key")
       }
     }
+    val removedKeysIterator = (tombstoneVals.keys ++ that.tombstoneVals.keys).iterator
+    var removedValues = Map.empty[A, B]
+    removedKeysIterator.foreach { key ⇒
+      (this.tombstoneVals.get(key), that.tombstoneVals.get(key)) match {
+        case (Some(thisValue), Some(thatValue)) ⇒
+          if (thisValue.getClass != thatValue.getClass) {
+            val errMsg = s"Wrong type for merging [$key] in [${getClass.getName}], existing type " +
+              s"[${thisValue.getClass.getName}], got [${thatValue.getClass.getName}]"
+            throw new IllegalArgumentException(errMsg)
+          }
+          val removedValue = thisValue.merge(thatValue.asInstanceOf[thisValue.T]).asInstanceOf[B]
+          removedValues = removedValues.updated(key, removedValue)
+        case (Some(thisValue), None) ⇒
+          removedValues = removedValues.updated(key, thisValue)
+        case (None, Some(thatValue)) ⇒
+          removedValues = removedValues.updated(key, thatValue)
+        case (None, None) ⇒ throw new IllegalStateException(s"missing value for $key")
+      }
+    }
+    // tombstone the values for removed keys
```

Comment:
Already `put` and `remove` propagate _delta_ for the `keys` `ORSet` but _full state_ for the `value`. Only `update` uses _delta-delta_ setup, and this is what results in anomaly, because we get only _delta_ of the second update but would lose the context.
I was thinking that perhaps it would be better to change existing semantics (and maybe a bit of an API). If every `DeltaReplicatedData` type had `clear` operation, it could be used to clear the value before removing the key. The relevant deltas would be propagated and we would be consistent in the operation semantics between full updates and deltas. However, this approach would have two drawbacks - for every key ever added we would keep forever at minimum its stub (consisting of key entry in the `keys` `ORSet` and the `vvector` of the `value`), and the semantics would be changed to "removal wins". So far the only way of getting rid of tombstones I figured out would be to not propagate the deltas of the `values` (but that wouldn't be a big win over not propagating deltas at all) or allow different behaviours for delta updates vs full updates (probably undesirable).